test/distributed/algorithms/test_join.py

```diff
@@ -0,0 +1,420 @@
+
+    @property
+    def world_size(self):
+        return WORLD_SIZE
+
+    @property
+    def process_group(self):
+        return dist.group.WORLD
+
+    def tearDown(self):
+        try:
+            dist.destroy_process_group()
+        except AssertionError:
+            pass
+        try:
+            os.remove(self.file_name)
+        except OSError:
+            pass
+
+    def dist_init(self, rank, world_size, backend=BACKEND):
+        store = dist.FileStore(self.file_name, world_size)
+        return dist.init_process_group(
+            backend=backend,
+            store=store,
+            rank=rank,
+            world_size=world_size
+        )
+
+    def construct_uneven_inputs(self, base, offset, device=None):
+        r"""
+        Returns uneven inputs: rank i gets ``base`` + i * ``offset``-many
+        inputs.
+        """
+        if device is None:
+            device = self.device
+        return [torch.zeros(1, device=device) for _ in range(base + self.rank * offset)]
+
+    def construct_even_inputs(self, base, device=None):
+        r"""Returns even inputs: each rank gets ``base``-many inputs."""
+        if device is None:
+            device = self.device
+        return [torch.zeros(1, device=device) for _ in range(base)]
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_single_joinable_main_hooks(self):
+        r"""Tests the main hooks of a single :class:`_Joinable`."""
+        self.dist_init(self.rank, self.world_size)
+        BASE_NUM_INPUTS = 3
+        OFFSET = 1
+
+        allreducer = AllReducer(self.device, self.process_group)
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        allreduce_total = 0
+        with _Join([allreducer]):
+            for _ in inputs:
+                allreduce_total += allreducer()
+
+        # Non-joined processes all-reduce a 1, so this rank's all-reduce total
+        # should be precisely equal to the total number of inputs processed
+        # before it joined
+        expected_total = self.world_size * BASE_NUM_INPUTS
+        # Rank i runs for i additional iterations
+        for num_joined in range(1, self.rank + 1):
+            expected_total += (self.world_size - num_joined) * OFFSET
+        self.assertEqual(allreduce_total, expected_total)
+
+    @require_n_gpus_for_nccl_backend(
+        WORLD_SIZE, BACKEND
+    )
+    def test_single_joinable_post_hooks(self):
+        r"""Tests the post-hooks of a single :class:`_Joinable`."""
+        self.dist_init(self.rank, self.world_size)
+        BASE_NUM_INPUTS = 3
+        OFFSET = 1
+
+        allreducer = AllReducer(self.device, self.process_group)
+        self.assertEqual(allreducer.post_hook_tensor.item(), BEFORE_CONSTANT)
+
+        inputs = self.construct_uneven_inputs(BASE_NUM_INPUTS, OFFSET)
+        with _Join([allreducer], run_post_hooks=True):
+            for _ in inputs:
+                allreducer()
```

Comment:
Would I be correct if I assume main hook would still run in this case? And compared to `test_single_joinable`, the only difference here is that `test_single_joinable_post_hooks` does not check main hook results? If so, does it make sense if we remove this test, as it seems `test_single_joinable` behaves exactly the same way and only differs in asserts?

Reply:
I have now made `test_single_joinable_post_hooks()` not run the main hooks.